### PR TITLE
Upgrade IronPython to v2.7.8 removing dependency on .NET 3.0

### PR DESCRIPTION
**Pull request by alfarok**
### Purpose
[QNTM-5357](https://jira.autodesk.com/browse/QNTM-5357) : Upgrade IronPython to 2.7.8.1
[QNTM-3509](https://jira.autodesk.com/browse/QNTM-3509) : Bundle IronPython w/ Core

This PR upgrades IronPython to v.2.7.8.1 using NuGet and removes any .NET 3.0 dependencies as per the request from corporate Delivery and Wingman teams and allows Revit web installs to be fully silent .  Additionally, it removes the committed IronPython binaries stored in the `extern` folder and the IronPython-2.7.3.msi from the `extra` folder and adds a NuGet reference to the IronPython Standard Library for v2.7.8.1.  When Dynamo is built this package is restored and copied to the bin directory.  The standard library is also now referenced by default.

```
# NO LONGER REQUIRED
import sys
sys.path.append(r"C:\Program Files (x86)\IronPython 2.7\Lib")
```

ironpythoninfo

### Installer Changes
[Installer PR](https://git.autodesk.com/Dynamo/DynamoForRevit/pull/154)

Revit 2020 no longer uses the Dynamo installer but it is still part of our CI/CD pipeline so changes will have to be made to the installer code to skip attempting to install Python using the previously included msi since IronPython will now be built and bundled with Core.

### Testing
All tests are passing on the self-serve.  I have manually verified in Sandbox and D4R in Revit 2020 that the correct version of IronPython is loaded from the `/bin` folder if a different version happens to be installed in the GAC (specifically version 2.7.3 which we used to install).  If an exact match occurs it does not seem possible to prevent loading from the GAC but this should be a non-issue.

ironpythoninfo_testgraph

Also added a test verifying the IronPython assemblies are only getting loaded from a single location.
image

### Declarations

Check these if you believe they are true

- [x] The code base is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [x] The level of testing this PR includes is appropriate
- [x] User facing strings, if any, are extracted into `*.resx` files
- [x] All tests pass using the self-service CI.
- [x] Snapshot of UI changes, if any.
- [x] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions), and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.

### Reviewers

@mjkkirschner 

### FYIs

@aparajit-pratap

### CODE REVIEW

**Comment by alfarok** [comment]
All unit tests pass with the except of 4.  This will be resolved when the version of IronPython is upgraded to 2.7.8 on the self-serve which the BRE team is already taking care of.  I will wait for that work to be completed before merging.

Additionally on Wednesday we are going to speak with the Revit team regarding how they plan to deliver IronPython in the installation.  See [QNTM-3509]( https://jira.autodesk.com/browse/QNTM-3509)

File: src/DynamoUtilities/DataMarshaler.cs [hunk 1]
```diff
@@ -76,14 +76,6 @@ public object Marshal(object obj)
 
             var targetType = obj.GetType();
 
-            // TODO: Remove this conversion after updating IronPython version in 2.1
-            // in which IronPython will support Int64
-            if (typeof (long) == targetType)
```
Comment (mjkkirschner) [inline src/DynamoUtilities/DataMarshaler.cs]: was there any change to the output marsheller?

Comment (alfarok) [reply src/DynamoUtilities/DataMarshaler.cs]: Not that I am aware of but @aparajit-pratap may have more knowledge here.  This was the original change I believe:

https://github.com/DynamoDS/Dynamo/pull/8602/files

**Comment by erfajo** [comment]
Will both 1.3.4 and 2.0.2 include the change to IronPython 2.7.8...

**Comment by alfarok** [comment]
@mjkkirschner The self-serve was just updated to use Python 2.7.8 so the 4 failing Python tests are now passing

**Comment by alfarok** [comment]
Note, a new self-serve was setup with an updated version of Python I am assuming for testing but the changes have not yet been done on the main Jenkins self-serve.

new:
https://master-15.jenkins.autodesk.com/view/DYN/job/DYN-DevCI_Self_Service_SDKChange/
usual:
https://master-15.jenkins.autodesk.com/view/DYN/job/DYN-DevCI_Self_Service/

**Comment by alfarok** [comment]
Also after some research I found the extern/IronPython .dlls need to be re-added and upgrade or the `Create Dynamo Core MSI` step will fail when building.  I will address this on Monday.

**Comment by mjkkirschner** [comment]
I'd like to consider not replacing these deps and removing everything from extern - just to figure out what that would take and if it should be part go this task or not.

**Comment by alfarok** [comment]
@mjkkirschner Think this is getting pretty close.  I am running the the self-serve one more time and looking into seeing how much work is involved in loading the python library references by default if they are available.

**Comment by alfarok** [comment]
@mjkkirschner I think this is ready for a final look when you are back.  I am just going to add a few more tests but everything is currently passing on the self-serve.

**Comment by alfarok** [comment]
Hey @ColinDayOrg @QilongTang, if either of you guys have a free moment could you take a look at this PR?

**Comment by jnealb** [comment]
@alfarok @mjkkirschner @ColinDayOrg @QilongTang Does the work-around for QNTM-3547 need to be revisited, eg removed?

**Comment by alfarok** [comment]
@jnealb I believe I removed the work-around referenced in that report and no issues marshalling int64 correctly without throwing an error like shown in QNTM-3547

image

**Comment by ColinDayOrg** [comment]
It is a little difficult to tell looking at raw project files like this, but it all seems good to me. :shipit:

**Comment by QilongTang** [comment]
@alfarok Looking at the workflow for any 1.3.X or 2.0.X user, do they update to the Dynamo 2.1 version (which does not remove old IronPython) then simply not use
```
import sys
sys.path.append(r"C:\Program Files (x86)\IronPython 2.7\Lib")
```
and everything still works? Can you clarify?

**Comment by alfarok** [comment]
@QilongTang the changes should not break any existing Python nodes but users could optionally remove those lines as they are no longer required.  My understanding is that you can a import module as many times as you want in one Python program, no matter what module it is. Every subsequent import after the first accesses the cached module instead of re-evaluating it.  

I do this by appending the new location of the standard library to the python engine directly and import `sys` by default.  One repercussion of this is that autocomplete for `sys` won't show unless imported again by the user since it is excluded from the template.  To prevent this we could add `import sys` to the template as well so all new instantiations of the node will include it.

 Having any additional versions of IronPython installed should not influence the proper version from being loaded either.

**Comment by QilongTang** [comment]
@alfarok Great. I just want to make sure.
1. Previous version user can still open legacy dyns with Python nodes with this change as long as they still have old Python installed.
2. We provide some way for users to adopt such version change.

If these two are met, I think this is safe change

**Comment by alfarok** [comment]
@QilongTang @mjkkirschner we could make this change to the template to be more clear if want.  Users will still have to include this statement if they want to get autocomplete when referencing 'sys'.  Let me know what you guys think.

image

**Comment by ColinDayOrg** [comment]
is there any downside to having the `include sys` line in the template? if not it seems pretty harmless and has a positive upside. (note that I don't know a lot about Python...)

**Comment by alfarok** [comment]
@ColinDayOrg Doesn't seem like there would be any issues other than it's technically a duplicate line but like I mentioned every subsequent import after the first accesses the cached module instead of re-evaluating it.  This is because previously existing python nodes in users graphs will not include any changes we make to the template so it has to be there for those scenarios.

**Comment by QilongTang** [comment]
@alfarok  I would support for updating the template

**Comment by mjkkirschner** [comment]
@alfarok I think it is important this PR includes a test that verifies python nodes which imported the libraries from the old path continue to work.

**Comment by QilongTang** [comment]
@alfarok @mjkkirschner Does any DynamoPythonTests cover running legacy graphs? I thought it should be covered already, no?

**Comment by mjkkirschner** [comment]
Not sure -do those tests consume standard python libraries that must be imported?

**Comment by alfarok** [comment]
@QilongTang @mjkkirschner I think some of this is covered here regarding loading the library and loading duplicate modules but I am adding a couple other verifications that I think will be useful
https://github.com/DynamoDS/Dynamo/blob/ec10f936824152e7dd7d6d019efdcda0d78a5264/test/Libraries/DynamoPythonTests/CodeCompletionTests.cs#L233

https://github.com/DynamoDS/Dynamo/blob/ec10f936824152e7dd7d6d019efdcda0d78a5264/test/Libraries/DynamoPythonTests/CodeCompletionTests.cs#L245

https://github.com/DynamoDS/Dynamo/blob/ec10f936824152e7dd7d6d019efdcda0d78a5264/test/Libraries/DynamoPythonTests/CodeCompletionTests.cs#L257

^BTW this new preview :thumbsup:

**Comment by alfarok** [comment]
@andydandy74 @mostaphaRoudsari @Amoursol @dimven Hey guys, the team is hoping to upgrade IronPython to v2.7.8 for Dynamo 2.1 and it will now be bundled with Dynamo Core to prevent conflicts when referencing the GAC.  Additionally we are now including the python standard library by default, which will be located in the root location where Dynamo is executing .

I don't believe these are breaking changes and legacy nodes should still function as expected, but we still have some additional QA testing to go through on our side.  I just wanted to give you a heads up as python package creators.  Feel free to bring up any concerns or tag others that may be interested.

**Comment by mjkkirschner** [comment]
LGTM

**Comment by Amoursol** [comment]
@alfarok If you want any more testing material, feel free to use any of the content on [the dynamoPython repo](https://github.com/Amoursol/dynamoPython)

**Comment by mostaphaRoudsari** [comment]
@alfarok, Beautiful! Thank you for the heads up. Is this only going to affect Dynamo 2.0 or is it a change for all versions of Dynamo?

**Comment by alfarok** [comment]
@mostaphaRoudsari This change will only be in 2.1 and forward so no changes to 1.3.X or 2.0.X.  Feel free to ping me if any additional issues or questions arise.  Hope all is well!

**Pull request merged**